packager/certcache/certcache.go

```diff
@@ -669,30 +750,64 @@ func (this *CertCache) updateCertIfNecessary() {
 		this.setCerts(certs)
 		return
 	}
-	if d >= time.Duration(CertRenewalInterval) {
+	if d >= time.Duration(certRenewalInterval) {
 		// Cert is still valid, don't do anything.
-	} else if d < time.Duration(CertRenewalInterval) {
-		// Cert is still valid, but we need to start process of requesting new cert.
-		log.Println("Warning: Current cert crossed threshold for renewal, attempting to renew.")
-		certs, err := this.certFetcher.FetchNewCert()
-		if err != nil {
-			log.Println("Error trying to fetch new certificates from CA: ", err)
-			return
+	} else if d < time.Duration(certRenewalInterval) {
+		this.renewedCertsMu.Lock()
+		defer this.renewedCertsMu.Unlock()
+
+		// Check if we already have a renewal cert waiting, fetch a new cert if not.
+		if this.renewedCerts == nil {
+			// Cert is still valid, but we need to start process of requesting new cert.
+			log.Println("Warning: Current cert crossed threshold for renewal, attempting to renew.")
+			certs, err := this.certFetcher.FetchNewCert()
+			if err != nil {
+				log.Println("Error trying to fetch new certificates from CA: ", err)
+				return
+			}
+			this.setNewCerts(certs)
+		} else {
+			var ocspUpdateAfter time.Time
+
+			ocsp, _, errorOcsp := this.readOCSP(true)
+			if errorOcsp != nil {
+				newOcsp := this.fetchOCSP(ocsp, this.renewedCerts, &ocspUpdateAfter, false)
```

Comment:
AFAICT, this is going to try to fetch the OCSP response for the new cert once a day, rather than using the backoff retry logic you added to readOCSP. Is that desired? Maybe that's OK for now, but can you add a TODO to revisit?

I realize it's difficult to use readOCSP here, since it's hard-coded to use this.certs and friends, rather than this.renewedCerts and friends. That makes me think two things:
1. The extraction of the retry logic from readOCSP would be useful.
2. We should bundle certName, certs, certsMu, ocspFile, ocspFilePath, ocspUpdateAfter, and ocspUpdateAfterMu into a new struct type, and then have two copies of that in certcache - one for current certs and one for new certs.

OK for future PR... TODO? :sheep: :grin:

Reply:
The logic for the certs is by design and is as follows:
updateCertIfNecessary() will be called:
1. On demand, during initialization or when GetLatestCert() is called and we already have an expired cert.
2. Every 'certCheckInterval' hours (currently every 24 hours).

The logic for retries in readOCSP is specific to just reading OCSP and retrying when we got a new cert because of Digicert's setup.

Inside updateCertIfNecessary, there's a case where we already got the renewalCert (waiting in this.renewedCerts) and in this case, we make sure we also get the OCSP by calling readOCSP with retries set to true.

I understand your suggestion of possibly reusing the readOCSP retry logic in this edge case, but given that the updateCertIfNecessary() calls are a day apart, chances are the fetchOCSP might fail the first time, but will succeed the next day.  Given that, it will still be nice-to-have the refactor of readOCSP so I added this as a TODO.